File: src/main/java/org/example/linkedlist/ReverseNodesInGroup.java
package org.example.linkedlist;

public class ReverseNodesInGroup {

    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode dummy = new ListNode(0, head);
        ListNode groupPrev = dummy;

        while (true) {
            ListNode kth = getKth(groupPrev, k);
            if (kth == null) break;

            ListNode groupNext = kth.next;
            ListNode prev = kth.next, curr = groupPrev.next, temp = null;

            while (curr != groupNext) {
                temp = curr.next;
                curr.next = prev;
                prev = curr;
                curr = temp;
            }

            temp = groupPrev.next;
            groupPrev.next = kth;
            groupPrev = temp;
        }

        return dummy.next;
    }

    private ListNode getKth(ListNode node, int k) {
        while (node != null && k > 0) {
            node = node.next;
            k -= 1;
        }

        return node;
    }
}

File: src/main/java/org/example/linkedlist/ReverseNodesInGroup.py
from typing import Optional

from org.example.linkedlist.ListNode import ListNode


class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        groupPrev = dummy

        while True:
            kth = self.getKth(groupPrev, k)
            if not kth: break

            groupNext = kth.next
            curr, prev = groupPrev.next, kth.next

            while curr != groupNext:
                temp = curr.next
                curr.next = prev
                prev = curr
                curr = temp

            temp = groupPrev.next
            groupPrev.next = kth
            groupPrev = temp

        return dummy.next

    def getKth(self, node: ListNode, k: int) -> Optional[ListNode]:
        while node and k > 0:
            node = node.next
            k -= 1
        return node
